packages/db-kit/src/utils/fetchExternal.ts

```diff
@@ -0,0 +1,192 @@
+
+  const { contracts } = await project.loadCompile({ result });
+
+  const contract = await findContract({
+    contracts,
+    contractName
+  });
+
+  await project.assignNames({
+    assignments: {
+      contracts: [contract.db.contract]
+    }
+  });
+
+  const { network } = await project.loadMigrate({
+    network: { name: config.network },
+    artifacts: [
+      {
+        ...Common.Shims.NewToLegacy.forContract(contract),
+        networks: {
+          [config.network_id]: {
+            address
+          }
+        }
+      }
+    ]
+  });
+
+  await project.assignNames({
+    assignments: {
+      networks: [network]
+    }
+  });
+}
+
+/**
+ * Accepts a TruffleConfig and returns a priority-sorted list of fetchers based
+ * on provided configuration and whether available fetchers support the current
+ * blockchain network.
+ * @dev HACK this currently only supports EtherscanFetcher
+ */
+export async function forTruffleConfig(options: {
+  config: TruffleConfig;
+}): Promise<Fetcher[]> {
+  const { config } = options;
+  const networkId = parseInt(config.network_id);
+
+  // respect optional `config.sourceFetchers`, used to specify explicit
+  // ordering for which fetchers to use.
+  //
+  // since this HACK-ily only supports Etherscan, this logic serves to respect
+  // configuration that explicitly excludes use of Etherscan.
+  const constructors: FetcherConstructor[] = config.sourceFetchers
+    ? config.sourceFetchers.map(sourceFetcherName => {
+        const Fetcher = Fetchers.find(
+          Fetcher => Fetcher.fetcherName === sourceFetcherName
+        );
+
+        if (!Fetcher) {
+          throw new Error(
+            `Unknown external source servce ${sourceFetcherName}.`
+          );
+        }
+        return Fetcher;
+      })
+    : Fetchers;
+
+  // construct fetchers for networkId
+  const fetchers = await Promise.all(
+    constructors
+      // HACK only support EtherscanFetcher right now
+      .filter(({ fetcherName }) => fetcherName === "etherscan")
+      .map(async (Fetcher: FetcherConstructor) => {
+        const options = config[Fetcher.fetcherName];
+
+        // don't fail on unsupported networks; that's expected behavior
+        try {
+          return await Fetcher.forNetworkId(networkId, options);
+        } catch (error) {
+          if (!(error instanceof InvalidNetworkError)) {
+            throw error;
+          }
+        }
+      })
+  );
+
+  return fetchers.filter((fetcher): fetcher is Fetcher => !!fetcher);
+}
+
+async function fetchAndCompile(options: {
+  config: TruffleConfig;
+  address: string;
+}): Promise<{
+  contractName: string;
+  result: WorkflowCompileResult;
+}> {
+  const { config, address } = options;
+  const fetchers = await forTruffleConfig({ config });
+
+  for (const fetcher of fetchers) {
+    //now comes all the hard parts!
+    //get our sources
+    let result;
+    try {
+      result = await fetcher.fetchSourcesForAddress(address);
```

Comment:
Can/should this be parallelized?

Reply:
Right now, the existence of this loop is just unnecessary boilerplate, since we only support one fetcher in db-kit right now (Etherscan).

In general though, this can't be parallelized, since the way truffle-config's `sourceFetchers` property works is that it specifies a priority-ordered list of fetchers. Not sure if this is a requirement that's worth changing (cc @haltman-at), but I posit that any such change now would be out of scope of this PR. Let me know if you concur @davidmurdoch! Thanks!